<a href="https://colab.research.google.com/github/ramkiran21/Codesoft/blob/main/Titanic_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Titanic dataset
Use the Titanic dataset to build a model that predicts whether a
passenger on the Titanic survived or not. This is a classic beginner
project with readily available data.
. The dataset typically used for this project contains information
about individual passengers, such as their age, gender, ticket
class, fare, cabin, and whether or not they survived.

In [5]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [29]:
# Importing the dataset
dataset = pd.read_csv('Titanic-Dataset.csv')
y = dataset['Survived'].values

# Drop columns not suitable for direct use
X_df = dataset.drop(['PassengerId', 'Survived', 'Name', 'Ticket', 'Cabin'], axis=1)
X = X_df.values

In [25]:
print('Missing values in the dataset:')
print(dataset.isnull().sum())

Missing values in the dataset:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [30]:
# Encoding the Categorical Variable 'Sex' (Index 1 in X)
from sklearn.preprocessing import LabelEncoder
labelencoder_X_sex = LabelEncoder()
X[:, 1] = labelencoder_X_sex.fit_transform(X[:, 1])

In [27]:
# labelencoder_X = LabelEncoder() # No longer needed for X in this cell as Ticket is dropped
# X[:, 8] = labelencoder_X.fit_transform(X[:, 8]) # Removed as Ticket column is dropped

# Encoding the Dependent Variable (Survived)
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

In [32]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [31]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
import numpy as np

# Impute missing 'Age' (index 2) and 'Fare' (index 5) with mean
imputer_mean = SimpleImputer(missing_values=np.nan, strategy='mean')
X[:, [2, 5]] = imputer_mean.fit_transform(X[:, [2, 5]])

# Impute missing 'Embarked' (index 6) with most frequent value
imputer_mode = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
X[:, 6] = imputer_mode.fit_transform(X[:, 6].reshape(-1, 1)).flatten()

# Encode 'Embarked'
labelencoder_X_embarked = LabelEncoder()
X[:, 6] = labelencoder_X_embarked.fit_transform(X[:, 6])

# Splitting the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [33]:
# Applying LDA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components=1)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)

# Logistic Regression
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

# Making the Confusion Matrix and checking accuracy
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:\n', cm)
print('Accuracy Score:', accuracy_score(y_test, y_pred))
print('\nClassification Report:\n', classification_report(y_test, y_pred))

Confusion Matrix:
 [[92 18]
 [20 49]]
Accuracy Score: 0.7877094972067039

Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.84      0.83       110
           1       0.73      0.71      0.72        69

    accuracy                           0.79       179
   macro avg       0.78      0.77      0.77       179
weighted avg       0.79      0.79      0.79       179



Based on our analysis, the final result for the Titanic survival prediction model is as follows:

Best Performing Model: Support Vector Machine (SVM) with an RBF kernel.
Final Accuracy: 79.89% on the test set.
Key Preprocessing: We handled missing values (imputing Age with mean and Embarked with mode), encoded categorical variables (Sex and Embarked), and applied feature scaling and Linear Discriminant Analysis (LDA) for dimensionality reduction.
While we started with a Logistic Regression baseline of 78.77%

In [36]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

# Dictionary of models to evaluate
models = {
    'Logistic Regression': LogisticRegression(random_state=0),
    'Random Forest': RandomForestClassifier(n_estimators=100, random_state=0),
    'SVM': SVC(kernel='rbf', random_state=0),
    'KNN': KNeighborsClassifier(n_neighbors=5)
}

results = {}

# Evaluate each model
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred_model = model.predict(X_test)
    acc = accuracy_score(y_test, y_pred_model)
    results[name] = acc
    print(f'{name} Accuracy: {acc:.4f}')

# Display results as a DataFrame
results_df = pd.DataFrame(list(results.items()), columns=['Model', 'Accuracy']).sort_values(by='Accuracy', ascending=False)
display(results_df)

Logistic Regression Accuracy: 0.7877
Random Forest Accuracy: 0.7542
SVM Accuracy: 0.7989
KNN Accuracy: 0.7654


,Model,Accuracy
2,SVM,0.798883
0,Logistic Regression,0.787709
3,KNN,0.765363
1,Random Forest,0.754190


In [35]:
from google.colab import sheets
sheet = sheets.InteractiveSheet(df=results_df)

https://docs.google.com/spreadsheets/d/1TtfHbp5NJYC5W7C2NiErPhXasjv43t3th_OBzGRe3II/edit#gid=0


Confusion Matrix:
 [[92 18]
 [20 49]]
Accuracy Score: 0.7877094972067039

Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.84      0.83       110
           1       0.73      0.71      0.72        69

    accuracy                           0.79       179
   macro avg       0.78      0.77      0.77       179
weighted avg       0.79      0.79      0.79       179

